core/src/external.rs

```diff
@@ -186,9 +202,23 @@ impl Value {
         }
     }
 
+    pub fn check_avm2_value(value: Avm2Value) -> Avm2Value {
+        match value {
+            Avm2Value::Undefined => {
+                if cfg!(not(target_family = "wasm")) {
```

Comment:
Hmm, can we simply return `undefined` on desktop as well? i.e. is the `null` case taken care of by the fact that there is no `ExternalInterface`, like you said?

Reply:
I'm wondering if we can instead do something like `ExternalInterface::available` in external.rs that returns `!self.providers.is_empty()`, and then the AVM's ExternalInterface.call` it returns `null` in that case. That way we don't need the `check_avm1_value`/`check_avm2_value` I believe?

Reply:
Looked into this a bit more, and the docs say (https://open-flash.github.io/mirrors/as2-language-reference/flash/external/ExternalInterface.html#call()):
`If the desired function is not available, the call returns null;`

So seems like the behavior should be:
 * If the external interface is not available or the function does not exist in the external interface, return `null`.
 * Otherwise, return what the function returns, which may be `undefined`.
 

I suspect in the browser, every function "exists" because it's JavaScript land, so this will always try to execute the function, which returns `undefined` naturally when it calls a non-existent function. I guess `JavascriptInterface::get_method` should return a dummy function/object that evaluates to `undefined` even when the method doesn't exist.

Reply:
I made the suggested changes as they definitely make sense. This works like a charm on web, returning "undefined" in the expected cases. Unfortunately, I'm also getting test failures. Here's one of them:

```diff
/// callWith() end
-undefined
+null
```
The Rust part of this test is defined here:
https://github.com/ruffle-rs/ruffle/blob/86395f194fd48cfbd46e01367125a3140fc8b5f7/tests/tests/external_interface/mod.rs#L26-L37

The ActionScript part is in [avm1/external_interface](https://github.com/ruffle-rs/ruffle/tree/master/tests/tests/swfs/avm1/external_interface):
```
function callWith(name, payload) {
   // (...)
   trace("/// callWith() end");
}
```

So, from my understanding, `callback.call(context, "callWith", ...)` runs and calls `callWith` which returns `undefined` as the function has no return value. This then causes `do_reentry` to return `ExternalValue::Undefined` instead of `ExternalValue::Null` and causes the test failure.

Reply:
I think the test output needs to be changed now, as it was testing for the incorrect behavior. Instead it should the test ouput should have `undefined` there. I verified in the browser that calling `player.foo()` from JS for a function that returns nothing will result in `undefined` in the Flash Player.